src/gnss.jl

```diff
@@ -0,0 +1,73 @@
+abstract type AbstractGNSSConstants end    
+abstract type AbstractGNSSData end
+
+@with_kw struct GNSSDecoderState{D <: AbstractGNSSData, C <: AbstractGNSSConstants, B <: Unsigned}
+    prn::Int
+    raw_buffer::B
+    buffer::B
+    raw_data::D
+    data::D
+    constants::C
+    num_bits_buffered::Int = 0
+    num_bits_after_valid_subframe::Union{Nothing, Int} = 0
+end
+
+function push_bit(state::GNSSDecoderState, current_bit)
+    GNSSDecoderState(
+        state,
+        raw_buffer = state.raw_buffer << true + (current_bit > 0),
+        num_bits_buffered = min(state.num_bits_buffered + 1, sizeof(state.raw_buffer) * 8)
+    )
+end
+
+function is_enough_buffered_bits_to_decode(state::GNSSDecoderState)
+    state.num_bits_buffered >=
+        state.constants.subframe_length + state.constants.preamble_length
+end
+
+calc_preamble_mask(state::GNSSDecoderState) = UInt(1) << UInt(state.constants.preamble_length) - UInt(1)
+
+function find_preamble(state::GNSSDecoderState)
+    state.raw_buffer & calc_preamble_mask(state) == state.constants.preamble &&
+        (state.raw_buffer >> state.constants.subframe_length) & calc_preamble_mask(state) == state.constants.preamble ||
+        state.raw_buffer & calc_preamble_mask(state) == ~state.constants.preamble &&
+        (state.raw_buffer >> state.constants.subframe_length) & calc_preamble_mask(state) == ~state.constants.preamble
+end
```

Comment:
I know it's safer to check for two preambles, but in my opinion, the checksum validation should catch all misdetected preambles so we could only check for one preamble and provide the data 8 bits earlier (for GPS L1)

Reply:
Yes, if we misdetect the preamble, several subsequent parity checks will probably fail. However, this would make the implementation more complex. How would you detect a misdetection? After 5 parity checks have failed? What if the parity check fails due to low SNR?